packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,221 @@
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+        require(feeProportion <= MAX_FEE, ERROR_INVALID_INPUT);
+
+        uint256 remainingEthAmount = msg.value;
+        if (feeProportion > 0 && feeRecipient != address(0x0)) {
+            // 1.5% is 150, allowing for 2 decimal precision, i.e 0.05% is 5
+            uint256 feeRecipientFeeAmount = safeDiv(safeMul(msg.value, feeProportion), PERCENTAGE_DENOMINATOR);
+            remainingEthAmount = safeSub(msg.value, feeRecipientFeeAmount);
+            // Transfer the fee to the fee recipient
+            feeRecipient.transfer(feeRecipientFeeAmount);
+        }
+
+        ETHER_TOKEN.deposit.value(remainingEthAmount)();
+        Exchange.FillResults memory fillTokensFillResult = marketBuyTokens(orders, signatures, feeOrders, feeSignatures, remainingEthAmount);
+        addFillResults(totalFillResult, fillTokensFillResult);
+        return totalFillResult;
+    }
+
+
+    function marketBuyTokens(
+        Order[] orders,
+        bytes[] signatures,
+        Order[] feeOrders,
+        bytes[] feeSignatures,
+        uint256 sellTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        uint256 takerTokenBalance = sellTokenAmount;
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1);
+
+        Exchange.FillResults memory tokensSellQuote = marketSellOrdersQuote(orders, sellTokenAmount);
+
+        if (tokensSellQuote.takerFeePaid > 0) {
+            // Fees are required for these orders
+            // Buy enough ZRX to cover the future market sell
+            Exchange.FillResults memory feeTokensResult =
+                buyFeeTokens(feeOrders, feeSignatures, tokensSellQuote.takerFeePaid);
+            takerTokenBalance = safeSub(takerTokenBalance, feeTokensResult.takerTokenFilledAmount);
+            totalFillResult.takerFeePaid = feeTokensResult.takerFeePaid;
+        }
+
+        // Make our market sell to buy the requested tokens with the remaining balance
+        Exchange.FillResults memory requestedTokensResult = EXCHANGE.marketSellOrders(orders, takerTokenBalance, signatures);
+        // Ensure the token abstraction was fair 
+        require(isAcceptableThreshold(sellTokenAmount, requestedTokensResult.takerTokenFilledAmount), ERROR_UNACCEPTABLE_THRESHOLD);
+        // Update our return FillResult with the market sell
+        addFillResults(totalFillResult, requestedTokensResult);
+        // Transfer all tokens to msg.sender
+        transferToken(makerTokenAddress, msg.sender, totalFillResult.makerTokenFilledAmount);
+        return totalFillResult;
+    }
+
+    function marketBuyNFT(
```

Comment:
Where is this used?

Reply:
This will be public once I've written the tests, since we have to transfer the NFT's individually.